std/stdio.d

```diff
@@ -4522,70 +4521,77 @@ Initialize with a message and an error code.
     }
 }
 
-extern(C) void std_stdio_static_this()
+// Undocummnted but public because the std* handles are aliasing it.
+ref File makeGlobal(alias handle)()
 {
-    static import core.stdc.stdio;
-    //Bind stdin, stdout, stderr
-    __gshared File.Impl stdinImpl;
-    stdinImpl.handle = core.stdc.stdio.stdin;
-    .stdin._p = &stdinImpl;
-    // stdout
-    __gshared File.Impl stdoutImpl;
-    stdoutImpl.handle = core.stdc.stdio.stdout;
-    .stdout._p = &stdoutImpl;
-    // stderr
-    __gshared File.Impl stderrImpl;
-    stderrImpl.handle = core.stdc.stdio.stderr;
-    .stderr._p = &stderrImpl;
-}
-
-//---------
-__gshared
-{
-    /** The standard input stream.
-        Bugs:
-            Due to $(LINK2 https://issues.dlang.org/show_bug.cgi?id=15768, bug 15768),
-            it is thread un-safe to reassign `stdin` to a different `File` instance
-            than the default.
-     */
-    File stdin;
-    ///
-    @safe unittest
+    static __gshared File.Impl impl;
+    static __gshared File result;
+    static shared uint initialized;
+    if (!initialized)
     {
-        // Read stdin, sort lines, write to stdout
-        import std.array : array;
-        import std.algorithm.sorting : sort;
-        import std.algorithm.mutation : copy;
-        import std.typecons : Yes;
-
-        void main() {
-            stdin                       // read from stdin
-            .byLineCopy(Yes.keepTerminator) // copying each line
-            .array()                    // convert to array of lines
-            .sort()                     // sort the lines
-            .copy(                      // copy output of .sort to an OutputRange
-                stdout.lockingTextWriter()); // the OutputRange
+        for (;;)
+        {
+            if (initialized > uint.max / 2)
+                break;
+            import core.atomic;
+            if (atomicOp!"+="(initialized, 1) == 1)
+            {
+                impl.handle = handle;
+                result._p = &impl;
+                atomicOp!"+="(initialized, uint.max / 2);
```

Comment:
No. `impl.handle`, `result._p` and `initialized` are different memory locations. A further load from result._p may get reordered with the store to impl.handle.